beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs

```diff
@@ -0,0 +1,472 @@
+//! Provides a mechanism which queues work for later processing.
+//!
+//! When the `beacon_processor::Worker` imports a block that is acceptably early (i.e., within the
+//! gossip propagation tolerance) it will send it to this queue where it will be placed in a
+//! `DelayQueue` until the slot arrives. Once the block has been determined to be ready, it will be
+//! sent back out on a channel to be processed by the `BeaconProcessor` again.
+//!
+//! There is the edge-case where the slot arrives before this queue manages to process it. In that
+//! case, the block will be sent off for immediate processing (skipping the `DelayQueue`).
+//!
+//! Aggregated and unaggregated attestations that failed verification due to referencing an unknown
+//! block will be re-queued until their block is imported, or until they expire.
+use super::MAX_SCHEDULED_WORK_QUEUE_LEN;
+use beacon_chain::{BeaconChainTypes, GossipVerifiedBlock, MAXIMUM_GOSSIP_CLOCK_DISPARITY};
+use eth2_libp2p::{MessageId, PeerId};
+use fnv::FnvHashMap;
+use futures::task::Poll;
+use futures::{Stream, StreamExt};
+use slog::{crit, debug, error, Logger};
+use slot_clock::SlotClock;
+use std::collections::{HashMap, HashSet};
+use std::pin::Pin;
+use std::task::Context;
+use std::time::Duration;
+use task_executor::TaskExecutor;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::error::Error as TimeError;
+use tokio_util::time::delay_queue::{DelayQueue, Key as DelayKey};
+use types::{Attestation, EthSpec, Hash256, SignedAggregateAndProof, SubnetId};
+
+const TASK_NAME: &str = "beacon_processor_reprocess_queue";
+
+/// Queue blocks for re-processing with an `ADDITIONAL_QUEUED_BLOCK_DELAY` after the slot starts.
+/// This is to account for any slight drift in the system clock.
+const ADDITIONAL_QUEUED_BLOCK_DELAY: Duration = Duration::from_millis(5);
+
+/// For how long to queue aggregated and unaggregated attestations for re-processing.
+pub const QUEUED_ATTESTATION_DELAY: Duration = Duration::from_secs(30);
```

Comment:
```suggestion
pub const QUEUED_ATTESTATION_DELAY: Duration = Duration::from_secs(12);
```

I'm not *sure* about this, but I'm thinking that one (mainnet) slot would allow us to have a smaller queue and encounter the scenario where an attestation "expires" less (i.e., the slot becomes to old to import).

Do you have any objections here?

Reply:
I thought a lot about this, without a clear solution. If I remember correctly the attn expires after an epoch and the 30secs are from a _reasonable time to complete the rpc request_. I'm fine changing it, since if it reaches the deadline, it means the block never arrived anyway

Reply:
Hmm, yeah there's no clear answer. Regarding a "reasonable time", I would think that time needs to be less than 12s since that's our synchronicity assumption in Eth2; if peers can't dish out blocks before their slots are finished then it becomes basically impossible to build a common chain.

For this reason, I'd lean towards the 12s. However, there might still be something I'm missing.